compiler/rustc_mir_transform/src/match_branches.rs

```diff
@@ -37,144 +151,350 @@ pub struct MatchBranchSimplification;
-                                kind: StatementKind::Assign(Box::new((*lhs, rhs))),
-                            }
+        let mut compare_types = Vec::new();
+        for (f, s) in iter::zip(first_stmts, second_stmts) {
+            let compare_type = match (&f.kind, &s.kind) {
+                // If two statements are exactly the same, we can optimize.
+                (f_s, s_s) if f_s == s_s => CompareType::Same(f_s),
+
+                // If two statements are assignments with the match values to the same place, we can optimize.
+                (
+                    StatementKind::Assign(box (lhs_f, Rvalue::Use(Operand::Constant(f_c)))),
+                    StatementKind::Assign(box (lhs_s, Rvalue::Use(Operand::Constant(s_c)))),
+                ) if lhs_f == lhs_s
+                    && f_c.const_.ty() == s_c.const_.ty()
+                    && f_c.const_.ty().is_integral() =>
+                {
+                    match (
+                        f_c.const_.try_eval_scalar_int(tcx, param_env),
+                        s_c.const_.try_eval_scalar_int(tcx, param_env),
+                    ) {
+                        (Some(f), Some(s)) if f == s => CompareType::Eq(lhs_f, f_c.const_.ty(), f),
+                        (Some(f), Some(s))
+                            if ((f_c.const_.ty().is_signed() || discr_ty.is_signed())
+                                && f.try_to_int(f.size()).unwrap()
+                                    == ScalarInt::try_from_uint(first_val, discr_size)
+                                        .unwrap()
+                                        .try_to_int(discr_size)
+                                        .unwrap()
+                                && s.try_to_int(s.size()).unwrap()
+                                    == ScalarInt::try_from_uint(second_val, discr_size)
+                                        .unwrap()
+                                        .try_to_int(discr_size)
+                                        .unwrap())
+                                || (Some(f) == ScalarInt::try_from_uint(first_val, f.size())
+                                    && Some(s)
+                                        == ScalarInt::try_from_uint(second_val, s.size())) =>
+                        {
+                            CompareType::Discr(
+                                lhs_f,
+                                f_c.const_.ty(),
+                                f_c.const_.ty().is_signed() || discr_ty.is_signed(),
+                            )
+                        }
+                        _ => {
+                            return false;
                         }
                     }
+                }
+
+                // Otherwise we cannot optimize. Try another block.
+                _ => return false,
+            };
+            compare_types.push(compare_type);
+        }
 
-                    _ => unreachable!(),
+        for (other_val, other_target) in iter {
```

Comment:
IIUC, this loop is some transposition of the previous one, considering that we already know which kind of comparison we want. Is that so?

Reply:
Yes. I added comments.